### PR TITLE
lapack/gonum: add Dpbtrf

### PR DESCRIPTION
**Pull request by vladimir-ch**
Please take a look.

### CODE REVIEW

File: lapack/gonum/dpbtf2.go [hunk 1]
```diff
@@ -89,16 +91,16 @@ func (Implementation) Dpbtf2(ul blas.Uplo, n, kd int, ab []float64, ldab int) (o
 		}
 		return true
 	}
+	// Compute the Cholesky factorization A = L * L^T.
 	for j := 0; j < n; j++ {
-		// Compute L(J,J) and test for non positive-definiteness.
+		// Compute L(j,j) and test for non-positive-definiteness.
```
Comment (kortschak) [inline lapack/gonum/dpbtf2.go]: We don't conventionally write hyphenation like this. I think we are wrong, so please leave it as is. (Just by way of a note).

File: lapack/gonum/ilaenv.go [hunk 2]
```diff
@@ -165,12 +165,12 @@ func (impl Implementation) Ilaenv(ispec int, name string, opts string, n1, n2, n
 				panic(badName)
 			case "TRF":
 				if sname {
-					if n4 <= 64 {
+					if n2 <= 64 {
```
Comment (kortschak) [inline lapack/gonum/ilaenv.go]: How did you find this?

File: lapack/testlapack/general.go [hunk 3]
```diff
@@ -1030,21 +1030,22 @@ func equalApproxSymmetric(a, b blas64.Symmetric, tol float64) bool {
...
 	a := make([]float64, n*n)
 	for i := 0; i < n; i++ {
-		for j := max(0, i-kb); j <= i; j++ {
+		for j := max(0, i-kd); j < i; j++ {
 			a[i*n+j] = rnd.NormFloat64()
 		}
-		a[i*n+i] = math.Abs(a[i*n+i])
-		// Add an extra amound to the diagonal in order to improve the condition number.
-		a[i*n+i] += 1.5 * rnd.Float64()
+		a[i*n+i] = 1 + math.Abs(rnd.NormFloat64())
```
Comment (kortschak) [inline lapack/testlapack/general.go]: Why this change?

File: lapack/gonum/dpbtrf.go [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+// license that can be found in the LICENSE file.
+
+package gonum
+
+import (
+	"gonum.org/v1/gonum/blas"
+	"gonum.org/v1/gonum/blas/blas64"
+)
+
+func (impl Implementation) Dpbtrf(uplo blas.Uplo, n, kd int, ab []float64, ldab int) (ok bool) {
```
Comment (kortschak) [inline lapack/gonum/dpbtrf.go]: Doc comment?

File: lapack/gonum/dpbtrf.go [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+		// storage.
+
+		// Process the band matrix one diagonal block at a time.
+		for i := 0; i < n; i += nb {
+			ib := min(nb, n-i)
+			// Factorize the diagonal block.
+			ok := impl.Dpotf2(uplo, ib, ab[i*ldab:], ldab-1)
+			if !ok {
+				return false
+
```
Comment (kortschak) [inline lapack/gonum/dpbtrf.go]: Delete blank line.

File: lapack/gonum/dpbtrf.go [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+			}
+			if i+ib >= n {
```
Comment (kortschak) [inline lapack/gonum/dpbtrf.go]: `>`? The Fortran has `i+ib.LE.n`. Is this a 1 versus 0?

File: lapack/gonum/dpbtrf.go [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+			//
+			// The numbers of rows and columns in the partitioning
+			// are ib, i2, i3 respectively. The blocks A12, A22 and
+			// A23 are empty if ib = kd. The upper triangle of A13
+			// lies outside the band.
+			i2 := min(kd-ib, n-i-ib)
+			if i2 > 0 {
+				// Update A12.
+				bi.Dtrsm(blas.Left, blas.Upper, blas.Trans, blas.NonUnit, ib, i2,
+					1, ab[i*ldab:], ldab-1, ab[i*ldab+ib:], ldab-1)
```
Comment (kortschak) [inline lapack/gonum/dpbtrf.go]: The absence of `kd` is row versus column major?

File: lapack/gonum/dpbtrf.go [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+	} else {
+		// Compute the Cholesky factorization of a symmetric band
+		// matrix, given the lower triangle of the matrix in band
+		// storage.
+
+		// Process the band matrix one diagonal block at a time.
+		for i := 0; i < n; i += nb {
+			ib := min(nb, n-i)
+			// Factorize the diagonal block.
+			ok := impl.Dpotf2(uplo, ib, ab[i*ldab+kd:], ldab-1)
```
Comment (kortschak) [inline lapack/gonum/dpbtrf.go]: And presence here and in the equivalent above for the same reason.

File: lapack/testlapack/dpbtrf.go [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+					case blas.Upper:
+						for i := 0; i < n; i++ {
+							for j := 0; j < min(kd+1, n-i); j++ {
+								diff = math.Max(diff, math.Abs(abFac[i*ldab+j]-ab[i*ldab+j]))
+							}
+						}
+					case blas.Lower:
+						for i := 0; i < n; i++ {
+							for j := max(0, i-kd); j <= i; j++ {
+								// diff = math.Max(diff, math.Abs(got[i*n+j]-abCopy[i*ldab+kd-i+j]))
```
Comment (kortschak) [inline lapack/testlapack/dpbtrf.go]: Is there here for a reason?

File: lapack/gonum/dpbtf2.go (hunk 1 above)
Comment (vladimir-ch) [reply lapack/gonum/dpbtf2.go]: Ok. The changes in this file are unnecessary but since I was going through it anyway, I thought I could do some minimal cleanup.

File: lapack/gonum/ilaenv.go (hunk 2 above)
Comment (vladimir-ch) [reply lapack/gonum/ilaenv.go]: Because the test was passing on the first try. Impossible, I thought. Then I found out that the blocked path was not taken because `Ilaenv` was always returning block size of 1 no matter how large the matrix/bandwidth was.

File: lapack/testlapack/general.go (hunk 3 above)
Comment (vladimir-ch) [reply lapack/testlapack/general.go]: For no particular reason. I was playing with different ways of generating the matrix and this was what stayed. I'll put the original code back although it doesn't really matter. This is again the problem of generating the test matrices. It would be nice to have the complete implementation of `DLATMS` which can also generate band matrices but `DLATMS` is ugly and complex and a lot of work.

File: lapack/gonum/dpbtf2.go (hunk 1 above)
Comment (kortschak) [reply lapack/gonum/dpbtf2.go]: Sorry. That was really unclear. Leave it how you have it.

File: lapack/gonum/dpbtrf.go (hunk 6 above)
Comment (vladimir-ch) [reply lapack/gonum/dpbtrf.go]: Yes, 1 vs 0 and the inverted condition. Another way to see it, `i+ib` is the _index_ where the next block would begin. Maybe it would be clearer to break out of the loop here and remove the end test in the for loop above. I just didn't want to have two apparent places to end the loop.

File: lapack/gonum/dpbtrf.go (hunk 7 above)
Comment (vladimir-ch) [reply lapack/gonum/dpbtrf.go]: Yes. There are two ways to translate Fortran banded to Go banded code:
1. The manual one: In Fortran, in case of Upper, the first index select the diagonal. `kd+1` is the main diagonal, `kd+1-1` is the first super-diagonal, etc. In general, `kd+1-d` is the d-th super-diagonal. So if we rewrite the row index in this form, we get our column index. For example, `1+IB = KD+1+IB-KD = KD+1-(KD-IB)`, so our column index would be `d = kd-ib`. To get the row index, we must subtract `d` from the Fortran column index. For example, `I+KD-d = I+KD-(KD-IB) = I+IB`. So the final index is `(i+ib)*ldab+kd-ib`.
2. The semi-manual one: In case of Upper, take Fortran code for Lower, swap row/column indices and convert to 0-based.

2 is of course less work but you have take care of sizes, sidedness, and transpose, which must match the actual uplo, not the uplo from which you are taking the indices. Overall, it's quite unpleasant and annoying.

File: lapack/testlapack/dpbtrf.go (hunk 9 above)
Comment (vladimir-ch) [reply lapack/testlapack/dpbtrf.go]: I don't know how this got here, maybe a github glitch? I had it in the code at some point but fixed before pushing. And I don't remember pushing to the branch after creating the PR and yet it is already marked as Outdated. Either way, it's not in the code now.

File: lapack/gonum/dpbtrf.go (hunk 4 above)
Comment (vladimir-ch) [reply lapack/gonum/dpbtrf.go]: Omg, how could I have forgotten this? :-) Added now.

File: lapack/testlapack/general.go (hunk 3 above)
Comment (vladimir-ch) [reply lapack/testlapack/general.go]: Done.

**Comment by vladimir-ch** [comment]
PTAL
